pingcap/tidb
# tidb-server: try best to get a meaningful IP

### What problem does this PR solve?

Issue Number: close #18399 

Problem Summary: While the newest tidb-server will set `cfg.AdvertiseAddr` to `cfg.Host`, it usually defaults to "0.0.0.0", which is not so useful. Now try best to get an IP, if there is no IP, or is the default "0.0.0.0". 

The code is injected into tidb-server.

### Check List

Tests 

- Unit test
- Manual test

Start TiDB, and run `curl http://127.0.0.1:xxx/info/all`, IP field should show your local area net IP address.

### Release note

- fix #18399, try best to get an meaningful IP
